### PR TITLE
Cli 5 tests prompts

### PR DESCRIPTION
**Pull request by tsvetomir-nedyalkov**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
     "chalk": "2.1.0",
     "config": "1.16.0",
     "console.table": "0.10.0",
-    "inquirer": "0.10.0",
+    "inquirer": "0.12.0",
```
Comment (DennyGD) [inline package.json]: README should be modified.  Using this version the CLI gives the impression that it should work fine in git bash where as it really doesn't.

File: test/TestsHelper.js [hunk 2]
```diff
@@ -117,8 +118,8 @@ TestsHelper.assertions = {
         return done(err);
       }
 
-      if (!expected || isEmpty(expected.flex)) {
-        const actualDoesNotContainData = isEmpty(actual) || isEmpty(actual.flex);
+      if (!expected) {
+        const actualDoesNotContainData = isEmpty(actual);
```
Comment (DennyGD) [inline test/TestsHelper.js]: Why the change?

File: test/TestsHelper.js [hunk 3]
```diff
@@ -183,7 +183,39 @@ TestsHelper.assertions = {
...
+    const result = {};
+    result[profileName] = { flex: serviceInfo };
+    return result;
+  },
+  assertFileContainsString(filePath, expectedString, done) {
+    fs.readFile(filePath, (err, data) => {
+      if (err) {
+        return done(err);
+      }
+      const fileContent = String.fromCharCode.apply(null, data);
```
Comment (DennyGD) [inline test/TestsHelper.js]: Isn't it OK to use just `String.fromCharCode(data)`?

File: test/TestsHelper.js [hunk 4]
```diff
@@ -410,8 +447,21 @@ TestsHelper.setup = {
...
+  },
+
+  startMockServer(options, cb, done) {
+    mockServer(options, (err, server) => {
+      if (err) {
+        return done(err);
+      }
+
+      cb(server);
+    });
```
Comment (DennyGD) [inline test/TestsHelper.js]: Having two callbacks is confusing the least. If an error doesn't occur, then `done` won't be called at all... Also, you're breaking standard node.js convention to pass error as first argument to callback and data as second.

File: test/integration/commands/flex/flex-init.test.js [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+const defaultEnvWithDebug = {
+  env: defaultEnv,
+  debug: fs.createWriteStream(outputFile)
+};
+
+const Prompt = {
+  selectAppOrOrg: /Would you like to select a service from a Kinvey app or org\? \(Use arrow keys\)/,
+  selectApp: /Which app would you like to use\?/,
+  selectService: /Which service would you like to use\?/,
+  selectOrganization: /Which organization would you like to use\? \(Use arrow keys\)/
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: These messages are available in `Constants`.

File: test/integration/commands/flex/flex-init.test.js [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+const appProjectFlex = assertions.buildExpectedProject(validDomains.app, fixtureApp.id, defaultService.id, defaultService.name);
+const expectedAppProject = assertions.buildExpectedProjectSetup(defaultProfileName, appProjectFlex);
+const orgProjectFlex = assertions.buildExpectedProject(validDomains.org, fixtureOrg.id, secondService.id, secondService.name);
+const expectedOrgProject = assertions.buildExpectedProjectSetup(defaultProfileName, orgProjectFlex);
+
+const secondValidProfileName = 'secondValidProfileName';
+const flexInitSuccessMessage = 'Saved configuration.';
+const notAuthenticatedMessage = 'You must be authenticated.';
+const notFoundProfileErrorMessage = 'Profile not found. Please verify profile name exists.';
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: Available in `Constants`.

File: test/integration/commands/flex/flex-init.test.js [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    .respond('\n')
+    .when(Prompt.selectService)
+    .respond('\n');
+
+  return sequence;
+};
+
+const buildProfileOptions = (baseOptions, profileName) => {
+  const profileOptions = cloneDeep(baseOptions);
+  profileOptions.push('--profile', profileName);
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: `Constants.AuthOptionsNames.PROFILE`

File: test/integration/commands/flex/flex-init.test.js [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+const secondValidProfileName = 'secondValidProfileName';
+const flexInitSuccessMessage = 'Saved configuration.';
+const notAuthenticatedMessage = 'You must be authenticated.';
+const notFoundProfileErrorMessage = 'Profile not found. Please verify profile name exists.';
+const notExistingProfileName = 'NotExistingProfile';
+
+const nodeCommand = 'node';
+const cliPath = path.join('bin', 'kinvey');
+const defaultFlexInitOptions = [cliPath, 'flex', 'init'];
+const profileEnvVarName = 'KINVEY_CLI_PROFILE';
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: `Constants.EnvironmentVariables.PROFILE`

File: test/integration/commands/flex/flex-init.test.js [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+        done();
+      });
+    } else {
+      done();
+    }
+  });
+
+  describe('App Level Services', () => {
+    it('with one not active valid profile should succeed', (done) => {
+      setup.createProfiles(defaultProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: And if an error occurs while creating?

File: test/integration/commands/flex/flex-init.test.js [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+          const sequence = buildSupposeFlexInitSequence(defaultFlexInitOptions, defaultEnvWithDebug);
+          runSupposeSequence(sequence, (error, exitCode) => {
+            assertions.assertSuccessfulFlexInitSequence(error, exitCode, expectedAppProject, outputFile, flexInitSuccessMessage, done);
+          });
+        }, done);
+      });
+    });
+
+    it('with one active valid profile should succeed', (done) => {
+      setup.setActiveProfile(defaultProfileName, true, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+          runSupposeSequence(sequence, (error, exitCode) => {
+            assertions.assertSuccessfulFlexInitSequence(error, exitCode, expectedAppProject, outputFile, flexInitSuccessMessage, done);
+          });
+        }, done);
+      });
+    });
+
+    it('from two profiles should use the active one', (done) => {
+      setup.createProfiles(secondValidProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        setup.setActiveProfile(defaultProfileName, true, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+            runSupposeSequence(sequence, (error, exitCode) => {
+              assertions.assertSuccessfulFlexInitSequence(error, exitCode, expectedAppProject, outputFile, flexInitSuccessMessage, done);
+            });
+          }, done);
+        });
+      });
+    });
+
+    it('should use the submitted profile as an option', (done) => {
+      setup.createProfiles(defaultProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        setup.setActiveProfile(secondValidProfileName, true, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+      });
+    });
+
+    it('should use the submitted profile as an environment variable', (done) => {
+      const envWithProfileVar = {
+        env: cloneDeep(defaultEnv),
+        debug: fs.createWriteStream(outputFile)
+      };
+      envWithProfileVar.env[profileEnvVarName] = defaultProfileName;
+      setup.createProfiles(defaultProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+        setup.setActiveProfile(secondValidProfileName, true, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+      });
+    });
+
+    it('submitted profile as an option should be used if there is another profile set as an environment variable', (done) => {
+      const envWithProfileVar = {
+        env: cloneDeep(defaultEnv),
+        debug: fs.createWriteStream(outputFile)
+      };
+      envWithProfileVar.env[profileEnvVarName] = secondValidProfileName;
+      setup.createProfiles(defaultProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+        setup.setActiveProfile(secondValidProfileName, true, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

Comment (DennyGD) [reply test/integration/commands/flex/flex-init.test.js]: By the way, I think that at least for some of the tests the profile creation step can be extracted to a `before` hook.

File: test/integration/commands/flex/flex-init.test.js [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+        runSupposeSequence(sequence, (error, exitCode) => {
+          expect(error.message).to.contain(notAuthenticatedMessage);
+          expect(exitCode).to.equal(1);
+          done();
+        });
+      }, done);
+    });
+
+    it('should return a not authenticated error message if there are more than one not active profiles', (done) => {
+      setup.createProfiles([secondValidProfileName, defaultProfileName], () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+            expect(error.message).to.contain(notAuthenticatedMessage);
+            expect(exitCode).to.equal(1);
+            done();
+          });
+        }, done);
+      });
+    });
+
+    it('should return a not found error message if a not existing profile is submitted as an option', (done) => {
+      setup.createProfiles(defaultProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/flex/flex-init.test.js [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+        setup.startMockServer(null, (server) => {
+          ms = server;
+          const profileOptions = buildProfileOptions(defaultFlexInitOptions, notExistingProfileName);
+          const sequence = buildSupposeFlexInitSequence(profileOptions, defaultEnvWithDebug);
+
+          runSupposeSequence(sequence, (error, exitCode) => {
+            expect(error.message).to.contain(notFoundProfileErrorMessage);
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: What if `error` is `null`?

File: test/integration/commands/flex/flex-init.test.js [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+        }, done);
+      });
+    });
+  });
+
+  describe('Organization Level Services', () => {
+    const orgOptions = { domainType: 'organizations', domainEntityId: fixtureOrg.id };
+
+    it('with one not active valid profile should succeed', (done) => {
+      setup.createProfiles(defaultProfileName, () => {
```
Comment (DennyGD) [inline test/integration/commands/flex/flex-init.test.js]: If an error occurs here?

File: test/integration/commands/init/init.test.js [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+
+const testsConfig = require('../../../TestsConfig');
+const { assertions, setup, getCreatedProfileMessage, runSupposeSequence } = require('../../../TestsHelper');
+
+const fixtureUser = require('./../../../fixtures/user.json');
+const mockServer = require('../../../mockServer');
+const path = require('path');
+
+const suppose = require('suppose');
+const fs = require('fs');
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: https://github.com/Kinvey/style-guides/blob/master/backend/es6.md#requires--order-by-scope

File: test/integration/commands/init/init.test.js [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+
+const outputFile = testsConfig.paths.supposeDebug;
+const existentUser = fixtureUser.existent;
+const nonExistentUser = fixtureUser.nonexistent;
+
+const initCommand = 'init';
+const invalidCredentialsMessage = 'Invalid credentials, please authenticate.';
+const invalidConfigUrlMessage = 'InvalidConfigUrl: The configuration URL is invalid. Please use a valid Kinvey instance name or URL.';
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: `Constants.Errors`

File: test/integration/commands/init/init.test.js [hunk 25]
```diff
@@ -0,0 +1,345 @@
...
+    host: testsConfig.host,
+    email: existentUser.email,
+    token: fixtureUser.token
+  };
+
+  const defaultProfileName = 'testProfile';
+  const expectedProfile = assertions.buildExpectedProfile(defaultProfileName, expectedValidUser.host, expectedValidUser.email, expectedValidUser.token);
+  const expectedProfiles = assertions.buildExpectedProfiles(expectedProfile);
+  const defaultExpectedSetup = assertions.buildExpectedGlobalSetup({}, expectedProfiles);
+  const valid2FAToken = '666666';
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: How about using the 2fa token from `fixtures/user`?

File: test/integration/commands/init/init.test.js [hunk 26]
```diff
@@ -0,0 +1,345 @@
...
+    env: invalidEnv,
+    debug: fs.createWriteStream(outputFile)
+  };
+
+  const Prompt = {
+    email: /\? E-mail \(email\) /,
+    password: /\? Password /,
+    instanceId: /\? Instance ID \(optional\) /,
+    profileName: /\? Profile name /,
+    twoFactorAuthToken: /Two-factor authentication token/
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: `Constants.PromptMessages`

File: test/integration/commands/init/init.test.js [hunk 27]
```diff
@@ -0,0 +1,345 @@
...
+  };
+
+  const invalidEmailMessageRegex = /Please enter a valid e-mail address/;
+  const provideNotEmptyStringRegex = /Please provide a non-empty string./;
+  const invalid2FATokenRegex = /Please enter a valid 2FA token \(6 digits\)./;
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: Constants.PromptMessages

File: test/integration/commands/init/init.test.js [hunk 28]
```diff
@@ -0,0 +1,345 @@
...
+
+  let ms = {};
+  const nodeCommand = 'node';
+  const cliPath = path.join('bin', 'kinvey');
+
+
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: multiple empty lines

File: test/integration/commands/init/init.test.js [hunk 29]
```diff
@@ -0,0 +1,345 @@
...
+          .respond('\n')
+          .when(Prompt.profileName)
+          .respond(`${defaultProfileName}\n`)
+          .when(Prompt.email)
+          .respond(`${existentUser.email}\n`)
+          .when(Prompt.password)
+          .respond(`${existentUser.password}\n`);
+
+        runSupposeSequence(sequence, (error, exitCode) => {
+          expect(error.message).to.contain(invalidCredentialsMessage);
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: If `error` is `null`?

File: test/integration/commands/init/init.test.js [hunk 30]
```diff
@@ -0,0 +1,345 @@
...
+          .respond(`${existentUser.email}\n`)
+          .when(Prompt.password)
+          .respond(`${existentUser.password}\n`)
+          .when(Prompt.instanceId)
+          .respond(`${invalidInstanceName}\n`)
+          .when(Prompt.profileName)
+          .respond(`${defaultProfileName}\n`);
+
+        runSupposeSequence(sequence, (error, exitCode) => {
+          expect(error.message).to.contain(invalidConfigUrlMessage);
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: If `error` is `null`?

File: test/integration/commands/init/init.test.js [hunk 31]
```diff
@@ -0,0 +1,345 @@
...
+        .respond(`${existentUser.password}\n`)
+        .when(Prompt.instanceId)
+        .respond('\n')
+        .when(Prompt.profileName)
+        .respond(`${defaultProfileName}\n`)
+        .when(Prompt.twoFactorAuthToken)
+        .respond(`${valid2FAToken}\n`);
+
+      runSupposeSequence(sequence, (error, exitCode) => {
+        expect(error.message).to.contain(requiredTwoFactorAuthMessage);
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: If `error` is `null`?

File: test/integration/commands/init/init.test.js [hunk 32]
```diff
@@ -0,0 +1,345 @@
...
+        .respond('\n')
+        .when(Prompt.profileName)
+        .respond(`${defaultProfileName}\n`)
+        .when(Prompt.twoFactorAuthToken)
+        .respond(`${invalid2FAToken}\n`)
+        .when(invalid2FATokenRegex)
+        .respond(`${valid2FAToken}\n`);
+
+      runSupposeSequence(sequence, (error, exitCode) => {
+        expect(error.message).to.contain(requiredTwoFactorAuthMessage);
```
Comment (DennyGD) [inline test/integration/commands/init/init.test.js]: If `error` is `null`?

File: package.json (hunk 1 above)
Comment (tsvetomir-nedyalkov) [reply package.json]: Evgeni will modify the README in master

File: test/TestsHelper.js (hunk 2 above)
Comment (tsvetomir-nedyalkov) [reply test/TestsHelper.js]: It seems that at some point this had broken some of the tests, but now they are passing, so I am reverting the change.

Comment (tsvetomir-nedyalkov) [reply test/TestsHelper.js]: Update: The change is needed, as I am passing as expected ``` {
  "flexListProfile": {
    "flex": {
      "domain": "app"
      "domainEntityId": "885f5d307afd4168bebca1a64f815c1e"
      "schemaVersion": 2
      "serviceId": "12378kdl2"
      "serviceName": "TestKinveyDatalink"
    }
  }
}``` and with the old version this will always execute ```expect(actualDoesNotContainData, `Setup at ${path} is empty.`).to.be.true;``` and the tests will always pass.

File: test/TestsHelper.js (hunk 4 above)
Comment (tsvetomir-nedyalkov) [reply test/TestsHelper.js]: Reverting to use directly the mockServer function

**Comment by DennyGD** [comment]
LGTM
